src/main/scala/mesosphere/marathon/core/volume/providers/DVDI.scala

```diff
@@ -0,0 +1,163 @@
+  }
+
+  def nameOf(vol: PersistentVolumeInfo): Option[String] = {
+    if (vol.providerName.isDefined && vol.name.isDefined) {
+      Some(vol.providerName.get + "::" + vol.name.get)
+    }
+    else None
+  }
+
+  // group-level validation for DVDI volumes: the same volume name may only be referenced by a single
+  // task instance across the entire cluster.
+  val groupValidation: Validator[Group] = new Validator[Group] {
+    override def apply(g: Group): Result = {
+      val groupViolations = g.apps.flatMap { app =>
+        val nameCounts = volumeNameCounts(app)
+        val internalNameViolations = {
+          nameCounts.filter(_._2 > 1).map{ e =>
+            RuleViolation(app.id, s"Requested volume ${e._1} is declared more than once within app ${app.id}", None)
+          }
+        }
+        val instancesViolation: Option[RuleViolation] =
+          if (app.instances > 1) Some(RuleViolation(app.id,
+            s"Number of instances is limited to 1 when declaring external volumes in app ${app.id}", None))
+          else None
+        val ruleViolations = app.container.toSet[Container].flatMap(DVDIProvider.this.collect).flatMap{ vol =>
+          val name = nameOf(vol.persistent)
+          if (name.isDefined) {
+            for {
+              otherApp <- g.transitiveApps.toList
+              if otherApp.id != app.id // do not compare to self
+              otherVol <- otherApp.container.toSet[Container].flatMap(DVDIProvider.this.collect)
+              otherName <- nameOf(otherVol.persistent)
+              if name.get == otherName
+            } yield RuleViolation(app.id,
+              s"Requested volume $name conflicts with a volume in app ${otherApp.id}", None)
+          }
+          else None
+        }
+        if (internalNameViolations.isEmpty && ruleViolations.isEmpty && instancesViolation.isEmpty) None
+        else Some(GroupViolation(app, "app contains conflicting volumes", None,
+          internalNameViolations.toSet[Violation] ++ instancesViolation.toSet ++ ruleViolations.toSet))
+      }
+      if (groupViolations.isEmpty) Success
+      else Failure(groupViolations.toSet)
+    }
+  }
+
+  def driversInUse(ct: Container): Set[String] =
+    DVDIProvider.this.collect(ct).flatMap(_.persistent.options.get(optionDriver)).toSet
+
+  /** @return a count of volume references-by-name within an app spec */
+  def volumeNameCounts(app: AppDefinition): Map[String, Int] =
+    app.container.toSet[Container].flatMap(DVDIProvider.this.collect).flatMap{ pv => nameOf(pv.persistent) }.
+      groupBy(identity).mapValues(_.size)
+
+  protected[providers] def modes(ct: Container): Set[Mode] =
+    DVDIProvider.this.collect(ct).map(_.mode).toSet
+
+  /** Only allow a single docker volume driver to be specified w/ the docker containerizer. */
+  val containerValidation: Validator[Container] = validator[Container] { ct =>
+    (ct.`type` is equalTo(ContainerInfo.Type.MESOS) and (modes(ct).each is equalTo(Mode.RW))) or (
```

Comment:
we may need to enforce `RW` for all containerizers; need to test docker containerizer support for RO